libbpf-tools/lsan.c

```diff
@@ -0,0 +1,1331 @@
+		}
+		rst = bpf_map_lookup_elem(sfd, &(curr->id), ip);
+		syms = syms_cache__get_syms(syms_cache, env.pid);
+		if (rst == 0 && syms != NULL) {
+			leak_printer(curr, ip, syms, "indirect");
+		}
+	}
+	free(ip);
+}
+
+static void delete_hash_lsan_hash(struct lsan_hash_t* hash) {
+	struct lsan_hash_t *curr = NULL;
+	struct lsan_hash_t *next = NULL;
+
+	HASH_ITER(hh, hash, curr, next) {
+		HASH_DEL(hash, curr);
+		free(curr);
+	}
+}
+
+static void delete_hash_report_info(struct report_info_t* hash) {
+	struct report_info_t *curr = NULL;
+	struct report_info_t *next = NULL;
+
+	HASH_ITER(hh, hash, curr, next) {
+		HASH_DEL(hash, curr);
+		free(curr);
+	}
+}
+
+static void delete_hash_root_region(struct root_region_t* hash) {
+	struct root_region_t *curr = NULL;
+	struct root_region_t *next = NULL;
+
+	HASH_ITER(hh, hash, curr, next) {
+		HASH_DEL(hash, curr);
+		free(curr);
+	}
+}
+
+static void empty_table()
+{
+	cvector_free(frontier);
+	cvector_free(key_table);
+	cvector_free(tids);
+	delete_hash_lsan_hash(copied);
+	delete_hash_report_info(direct);
+	delete_hash_report_info(indirect);
+	delete_hash_root_region(certain);
+	delete_hash_root_region(uncertain);
+	frontier = NULL;
+	key_table = NULL;
+	tids = NULL;
+	copied = NULL;
+	direct = NULL;
+	indirect = NULL;
+	certain = NULL;
+	uncertain = NULL;
+}
+
+static void for_each_tid_ptrace(enum __ptrace_request request)
+{
+	int i = 0;
+	long ret = 0;
+
+	if (!env.stop_the_world)
+		return;
+
+	for (i = 0; i < cvector_size(tids); ++i) {
+		ret = ptrace(request, tids[i], NULL, NULL);
+		if (ret == -1) {
+			p_warn("ptrace failed to request %d, reason: %s", request, strerror(errno));
+			p_warn("May failed to stop tid: %d, could cause false alarms", tids[i]);
+		}
+	}
+}
+
+static void for_each_tid_waitpid()
+{
+	int i = 0;
+	int status = 0;
+	pid_t ret = 0;
+
+	if (!env.stop_the_world)
+		return;
+
+	for (i = 0; i < cvector_size(tids); ++i) {
+		ret = waitpid(tids[i], &status, __WALL);
+		if (ret == -1) {
+			p_warn("waitpid failed, reason: %s", strerror(errno));
+			p_warn("May failed to stop tid: %d, could cause false alarms", tids[i]);
+		}
+	}
+}
+
+static int do_leak_check(struct syms_cache *syms_cache)
+{
+	int ret = 0;
+
+	get_tids();
+	for_each_tid_ptrace(PTRACE_SEIZE);
+	for_each_tid_ptrace(PTRACE_INTERRUPT);
+	for_each_tid_waitpid();
+	save_roots();
+	ret = read_table();
+	if (ret < 0) {
+		p_warn("Failed to read_table, retry after %d seconds", env.interval);
+		for_each_tid_ptrace(PTRACE_DETACH);
+		return 0;
+	}
+	classify_all_chunks();
+	for_each_chunk(collect_leaks_cb);
+	for_each_tid_ptrace(PTRACE_DETACH);
+	report_leaks(syms_cache);
+	return 0;
+}
+
+int main(int argc, char **argv)
+{
+	struct syms_cache *syms_cache = NULL;
+	int err = 0;
+	char path[PATH_MAX] = {0, };
+	FILE *fp_suppression = NULL;
+	char line[LINE_MAX] = {0, };
+	int i = 0;
+	char *ptr = NULL;
+	char *str = NULL;
+
+	set_log_level(INFO);
+#ifdef __arm__
+	p_err("lsan is not supported on arm32 systems");
+	return 0;
```

Comment:
Are you meaning that it would be better to error at compile time?
Sound reasonable, but I think I should check it with my coworkers.
Since compile error may affect build system in my company.